optimizely/config_manager.py

```diff
@@ -0,0 +1,198 @@
+    if url is None:
+      return url_template.format(sdk_key=sdk_key)
+
+    return url
+
+  def get_update_interval(self, update_interval):
+    """ Helper method to determine frequency at which datafile has to be polled and ProjectConfig updated.
+
+    Args:
+      update_interval: Time in seconds optionally sent in by the user.
+    """
+    polling_interval = update_interval or enums.ConfigManager.DEFAULT_UPDATE_INTERVAL
+
+    # If polling interval is less than minimum allowed interval then set it to default update interval.
+    if update_interval < enums.ConfigManager.MIN_UPDATE_INTERVAL:
+      self.logger.debug('Invalid update_interval {} provided. Defaulting to {}'.format(
+        update_interval,
+        enums.ConfigManager.DEFAULT_UPDATE_INTERVAL)
+      )
+      polling_interval = enums.ConfigManager.DEFAULT_UPDATE_INTERVAL
+
+    return polling_interval
+
+  def set_last_modified(self, response):
+    """ Looks up and sets last modified time based on Last-Modified header in the response.
+
+     Args:
+       response: requests.Response
+     """
+    self.last_modified = response.headers.get(enums.HTTPHeaders.LAST_MODIFIED)
+
+  def set_config(self, response):
+    """ Looks up and sets datafile and config based on response body.
+
+     Args:
+       response: requests.Response
+     """
+    try:
+      response.raise_for_status()
+    except requests_exceptions.HTTPError as err:
+      self.logger.error('Fetching datafile from {} failed. Error: {}'.format(self.datafile_url, str(err)))
+      return
+
+    # Leave datafile and config unchanged if it has not been modified.
+    if response.status_code == http.HTTPStatus.NOT_MODIFIED:
+      self.logger.debug('Not updating config as datafile has not updated since {}.'.format(self.last_modified))
+      return
+
+    self.set_last_modified(response)
+    # TODO(ali): Add validation here to make sure that we do not update datafile and config if not a valid datafile.
+    self._datafile = response.text
+    # TODO(ali): Add notification listener.
+    self._config = project_config.ProjectConfig(self._datafile, self.logger, self.error_handler)
+    self.logger.info('Received new datafile and updated config.')
+
+  def get_config(self):
+    """ Returns instance of ProjectConfig.
+
+    Returns:
+      ProjectConfig.
+    """
+    return self._config
+
+  def fetch_datafile(self):
+    """ Fetch datafile and set ProjectConfig. """
+
+    request_headers = {}
+    if self.last_modified:
+      request_headers[enums.HTTPHeaders.IF_MODIFIED_SINCE] = self.last_modified
+
+    response = requests.get(self.datafile_url, headers=request_headers)
+    self.set_config(response)
+
+  def _run(self):
+    """ Triggered as part of the thread which fetches the datafile and sleeps until next update interval. """
+    while self.is_running:
+      self.fetch_datafile()
+      time.sleep(self.update_interval)
+
+  def start(self):
```

Comment:
I think both `start` and `stop` should be methods on the base manager because I can see it being implemented in other types of implementations like push-based ones. In the static one, they can just be No-Op. I can see the Optimizely SDK, which will use this manager, would want to call `start`/`close` on this irrespective of the impl used.

Reply:
I was going by the interface we arrived on when implementing the Java SDK. I think at its core, the config manager should be responsible for returning the config through a `get_config` method. 
Building start/stop should not be necessary to be able to use it. In the current work, we can actually start the thread in constructor and not worry about stop as when it goes out of scope, the thread will be stopped anyway.